### PR TITLE
Add RadiusNeighborhood as a PairwiseDistancesReduction 

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference Issues/PRs

Built on top of scikit-learn#20254

#### What does this implement/fix? Explain your changes.

Add a new reduction for querying neighbors using a radius (similarly to [`RadiusNeighborsMixin._radius_neighbors_reduce_func`](https://github.com/scikit-learn/scikit-learn/blob/d386fe11a58ae910e988a17c02276964b917bce1/sklearn/neighbors/_base.py#L926)) using `std::vector<std::vector> >`, see [this snippet](https://gist.github.com/jjerphan/6b90a9691e67444adc7a8a658747ebe1).

### CODE REVIEW

**Comment by ogrisel** [comment]
Could you check that this does not cause a memory leak by calling RadiusNeighbors' query method in a loop, discarding the results each time and monitorying `psutil.Process().memory_info().rss` every 100 iterations of the loop for instance?

Try a big enough test set with k large enough such that the resulting neighbors data is around 1MB in memory for instance. The training set does not need to be large.

File: sklearn/neighbors/tests/test_neighbors.py [hunk 1]
```diff
@@ -1909,3 +1910,38 @@ def test_fast_sqeuclidean_translation_invariance(
...
+        rng.rand(int(n * d / ratio_train_test)).astype(dtype).reshape((-1, d)) * spread
+    )
+
+    rn = RadiusNeighborhood.get_for(X=X_test, Y=X_train, radius=radius, metric=metric)
+    dists, indices = rn.compute(return_distance=True, sort_results=True)
+    nn = NearestNeighbors(radius=radius, metric=metric)
+    nn.fit(X_train)
+    reference_dists, references_indices = nn.radius_neighbors(
+        X_test, return_distance=True, sort_results=True
+    )
```
Comment (ogrisel) [inline sklearn/neighbors/tests/test_neighbors.py:1943]: At some point in the future,  `NearestNeighbors.radius_neighbors` will rely on this new `RadiusNeighborhood` Cython code and therefore this test will be meaning less.

Would it be possible to instead compare the consistency to naive Python code that uses scipy's `cdist` + basic numpy operations for computing the radius neighbors?

Comment (ogrisel) [reply sklearn/neighbors/tests/test_neighbors.py:1943]: Actually, I just saw:

```python
# Temporary consistency check
```

and indeed the existing scikit-learn test suite for `NearestNeighbors.radius_neighbors` is probably enough. So I am fine with keeping this as a temporary test to be removed once `NearestNeighbors` is updated to leverage `RadiusNeighborhood` internally. Maybe add a comment saying so to make that more explicit.

File: sklearn/neighbors/tests/test_neighbors.py [hunk 2]
```diff
@@ -1909,3 +1910,38 @@ def test_fast_sqeuclidean_translation_invariance(
 
     assert_allclose(reference_dist, dist)
     assert_array_equal(reference_nns, nns)
+
+
+@pytest.mark.parametrize("n", [10 ** i for i in [3, 4]])
+@pytest.mark.parametrize("d", [2])
```
Comment (ogrisel) [inline sklearn/neighbors/tests/test_neighbors.py:1916]: nitpick: better use `n_samples` / `n_features` consistently :)

File: sklearn/metrics/_parallel_reductions.pyx [hunk 3]
```diff
@@ -721,3 +748,154 @@ cdef class FastSquaredEuclideanArgKmin(ArgKmin):
...
+
+    def _finalise_compute(self,
+           bint return_distance
+    ):
+        if return_distance:
+            np_arrays_indices = []
+            np_arrays_distances = []
+
+            for i in range(self.n_X):
+                np_arrays_distances.append(buffer_to_numpy_array(deref(self.neigh_distances)[i].data(),
```
Comment (ogrisel) [inline sklearn/metrics/_parallel_reductions.pyx:888]: You could even write a helper named `nested_vectors_to_numpy_arrays` to loop over the outer vectors and to also take care of the deref + `.data()`.

Comment (ogrisel) [reply sklearn/metrics/_parallel_reductions.pyx:888]:
```python
if return_distance:
    return (
        nested_vectors_to_numpy_arrays(self.neigh_distances),
        nested_vectors_to_numpy_arrays(self.neigh_indices),
    )
else:
   return nested_vectors_to_numpy_arrays(self.neigh_indices)
```

Comment (ogrisel) [reply sklearn/metrics/_parallel_reductions.pyx:888]: And you also need to free `self.neigh_distances`  if `return_distance` is `False`.

File: sklearn/metrics/_parallel_reductions.pyx [hunk 4]
```diff
@@ -721,3 +748,154 @@ cdef class FastSquaredEuclideanArgKmin(ArgKmin):
...
+        # as the ownership will be transferred to
+        # numpy arrays.
+        #
+        # TODO: Find a proper way to handle this
+        # It's "OK-ish" as numpy arrays are
+        # then responsible for their buffer
+        # lifetime which consist of most of the
+        # vectors actual data (residual metadata
+        # exist, don't account but won't be deleted).
+        # Still, meh.
```
Comment (ogrisel) [inline sklearn/metrics/_parallel_reductions.pyx:789]: Please use your 80 char limit in the comments :)

If you use VS Code you can use https://marketplace.visualstudio.com/items?itemName=konstantin.wrapSelection

**Review by ogrisel — COMMENTED** [review]
Exciting! Here is a few feedback:

File: sklearn/neighbors/tests/test_neighbors.py (hunk 1 above)
Comment (jjerphan) [reply sklearn/neighbors/tests/test_neighbors.py:1943]: Yes, I should have named it "Temporary transitional test".

I will make it more explicit.

File: sklearn/metrics/_parallel_reductions.pyx (hunk 3 above)
Comment (jjerphan) [reply sklearn/metrics/_parallel_reductions.pyx:888]: > You could even write a helper named nested_vectors_to_numpy_arrays to loop over the outer vectors and to also take care of the deref + .data().

This is what I did at first, but I got some weird problem with a missing symbol in the generated .so file (`_ZSt28__throw_bad_array_new_lengthv`) when typing it. This is related to numpy in some way, but I haven't yet be able to diagnose why. 

So yes, this is _highly ugly_ but as it worked and as this issue was capilotracting me, I decided to leave it as is for now to see if it was possible at least to have something working.

I will definitely change it. :+1:

**Comment by jjerphan** [comment]
Thanks for the review! Note that as of now, it's still WIP with some not so nice tactical advances to see if it can be functional before making it clean without too much difficulty, shall this be made functional.

**Comment by jjerphan** [comment]
**Quick report regarding memory usage:**

As of b8fe6e1c8ed38cf6a7d25d4d661dfaae389ad6e0, It looks like some data isn't deleted after calls to `RadiusNeighborhood.compute` solely.
This data size is proportional to the number of query vectors.

n_calls=100_radius=10000_return_distance=False
n_calls=100_radius=0_return_distance=True
n_calls=100_radius=0_return_distance=False
n_calls=100_radius=10000_return_distance=True

It looks like numpy arrays do not free their buffer, one can attest this applying this patch:
```patch
diff --git a/sklearn/metrics/_parallel_reductions.pyx b/sklearn/metrics/_parallel_reductions.pyx
index 57f3146664..95b60b2cd6 100644
--- a/sklearn/metrics/_parallel_reductions.pyx
+++ b/sklearn/metrics/_parallel_reductions.pyx
@@ -1052,3 +1052,6 @@ cdef class RadiusNeighborhood(PairwiseDistancesReduction):
         # by a numpy array then.
         del self.neigh_distances
         return _coerce_vectors_to_np_nd_arrays(self.neigh_indices)
+
+    def indices(self):
+        return np.asarray(deref(self.neigh_indices))
\ No newline at end of file
```

and seeing that `RadiusNeighborhood.indices` still returns a valid results after having `del` the numpy arrays:
```python
In [1]: %run script_given_bellow.py
In [2]: dists, indices = rn.compute(return_distance=return_distance, sort_results=True, strategy="parallel_
   ...: on_Y")

In [3]: indices
Out[3]: 
array([array([4, 2, 5, 3, 0, 1, 6, 7, 8, 9]),
       array([6, 2, 5, 9, 1, 4, 3, 0, 8, 7]),
       array([4, 0, 9, 1, 3, 2, 7, 5, 6, 8]), ...,
       array([6, 8, 0, 3, 7, 9, 4, 1, 2, 5]),
       array([6, 1, 4, 2, 0, 7, 8, 3, 5, 9]),
       array([8, 9, 1, 6, 5, 3, 4, 7, 0, 2])], dtype=object)

In [4]: del indices

In [5]: rn.indices()
Out[5]: 
array([[4, 2, 5, ..., 7, 8, 9],
       [6, 2, 5, ..., 0, 8, 7],
       [4, 0, 9, ..., 5, 6, 8],
       ...,
       [6, 8, 0, ..., 1, 2, 5],
       [6, 1, 4, ..., 3, 5, 9],
       [8, 9, 1, ..., 7, 0, 2]])
```


Probably numpy's memory allocator is different and proper handling is needed as suggested by [this response](https://stackoverflow.com/a/55959886).

<details>
<summary> Script </summary>

```python
from sklearn.metrics._parallel_reductions import RadiusNeighborhood
import numpy as np
import psutil
import pandas as pd
import gc
import matplotlib.pyplot as plt

if __name__ == "__main__":
    memory_info = []
    mem_info = psutil.Process().memory_full_info()
    memory_info.append(mem_info)
    rng = np.random.RandomState(0)
    X = rng.rand(10000, 100)
    Y = rng.rand(10, 100)
    mem_info = psutil.Process().memory_full_info()
    memory_info.append(mem_info)

    n_calls = 100
    radius = 10000
    return_distance = True

    for i in range(n_calls):
        rn = RadiusNeighborhood.get_for(X, Y, radius=radius)
        # Results are discarded
        rn.compute(return_distance=return_distance, sort_results=True, strategy="parallel_on_Y")
        mem_info = psutil.Process().memory_full_info()
        print(f"Iteration {i} / {n_calls}", mem_info)
        memory_info.append(mem_info)

    gc.collect()
    mem_info = psutil.Process().memory_full_info()
    memory_info.append(mem_info)

    title = "Memory usage on calls to RadiusNeighborhood.compute"
    title += f"\n X.shape={X.shape} -- Y.shape={Y.shape} " \
             f"radius={radius} -- return_distance={return_distance}"

    df = pd.DataFrame(data=memory_info)
    df[["data", "rss", "uss"]].plot(title=title,
                                    xlabel="Number of calls",
                                    ylabel="Size of segment (in bytes)",
                                    figsize=(16, 9), fontsize=10)
    plt.savefig(f"/tmp/n_calls={n_calls}_radius={radius}_return_distance={return_distance}.png")
    plt.show()

```

</details>
